docs/csharp/tutorials/pattern-matching.md

```diff
@@ -0,0 +1,386 @@
+
+```csharp
+private static bool IsWeekDay(DateTime timeOfToll) =>
+    timeOfToll.DayOfWeek switch
+    {
+        DayOfWeek.Saturday,  => false,
+        DayOfWeek.Sunday => false,
+        _ => true
+    };
+```
+
+You can make a further simplification to this expression by combining the two values into one arm:
+
+[!code-csharp[IsWeekDay](../../../samples/csharp/tutorials/patterns/finished/toll-calculator/TollCalculator.cs#IsWeekDay)]
+
+Next, add a similar function to categorize the time into the blocks:
+
+[!code-csharp[GetTimeBand](../../../samples/csharp/tutorials/patterns/finished/toll-calculator/TollCalculator.cs#GetTimeBand)]
+
+The previous method doesn't use pattern matching. It's clearer using a familiar cascade of `if` statements. You do add a private `enum` to convert each range of time to a discrete value.
+
+After you create those methods, you can use another `switch` expression with the **tuple pattern** to calculate the pricing premium. You could build a `switch` expression with all 16 arms:
+
+[!code-csharp[FullTuplePattern](../../../samples/csharp/tutorials/patterns/finished/toll-calculator/TollCalculator.cs#TuplePatternOne)]
+
+The above code works, but it can be simplified. All eight combinations for the weekend have the same toll. You can replace all eight with the following one line:
+
+```csharp
+(false, _, _) => 1.0m,
+```
+
+Both inbound and outbound traffic have the same multiplier during the weekday daytime and overnight hours. Those four switch arms can be replaced with the follow two lines:
+
+```csharp
+(true, TimeBand.Overnight, _) => 0.75m,
+(true, TimeBand.Daytime, _) => 1.5m,
+```
+
+The code should look like the following code after those two changes:
+
+```csharp
+public decimal PeakTimePremium(DateTime timeOfToll, bool inbound) =>
+    (IsWeekDay(timeOfToll), GetTimeBand(timeOfToll), inbound) switch
+    {
+        (true, TimeBand.MorningRush, true) => 2.00m,
+        (true, TimeBand.MorningRush, false) => 1.00m,
+        (true, TimeBand.Daytime, _) => 1.50m,
+        (true, TimeBand.EveningRush, true) => 1.00m,
+        (true, TimeBand.EveningRush, false) => 2.00m,
+        (true, TimeBand.Overnight, _) => 0.75m,
+        (false, _, _) => 1.00m,
+    };
+```
+
+Finally, you can remove the two rush hour times that pay the regular price. Once you remove those arms, you can replace the `false` with a discard (`_`) in the final switch arm. You'll have the following finished method:
+
+[!code-csharp[SimplifiedTuplePattern](../../../samples/csharp/tutorials/patterns/finished/toll-calculator/TollCalculator.cs#FinalTuplePattern)]
+
+This example highlights one of the advantages of pattern matching. The pattern branches are evaluated in order. If you rearrange them so that an earlier branch handles one of your later cases, the compiler warns you. Those language rules made it easier to do the preceding simplifications with confidence that the code didn't change.
```

Comment:
There are many ways to screw up where the compiler won't warn you. I think that things are evaluated in order should be early in the article. Possibly mention that unreachable code is still a thing for switch expressions. But overall this paragraph doesn't have the same weight as the rest of the article, imo.

Reply:
I did introduce the ordering concept earlier. Other than that, I left this unchanged